### PR TITLE
(WIP) Add nonprofit profile edit functionality - tests unfinished

### PR DESCRIPTION
**Pull request by ianroberts131**
Eliot,

I'm having a bit of trouble with the tests ( I know, wrong order :) ), but did quite a bit on this branch so wanted to commit.

### CODE REVIEW

File: app/controllers/nonprofits_controller.rb [hunk 1]
```diff
@@ -1,5 +1,39 @@
 class NonprofitsController < ApplicationController
+  
+  before_action :authorize_manager, only: [:edit, :update]
+  
   def show
     @nonprofit = Nonprofit.find(params[:id])
   end
+  
+  def edit
+    @nonprofit = Nonprofit.find(params[:id])
```
Comment (eliotsykes) [inline app/controllers/nonprofits_controller.rb]: @ianroberts131 Good to see the authorization protection for the `edit` and `update` actions. Consider adding a further layer of security by scoping the `find` to the `current_user`:

```rb
# Can be used on lines 10 and 14:
@nonprofit = current_user.nonprofits.find(params[:id])
```

File: app/views/nonprofits/edit.html.erb [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+    
+    <%= form_for(@nonprofit, url: nonprofit_path, html: { method: :put } ) do |f| %>
+      <div class="form-group">
+        <%= f.label :name %><br />
+        <%= f.text_field :name, autofocus: true, class: "form-control", placeholder: "Update nonprofit name" %>
+      </div>
+      
+      <div class="form-group">
+        <%= f.label :email %><br />
+        <%= f.text_field :email, class: "form-control", placeholder: "Update website" %>
```
Comment (eliotsykes) [inline app/views/nonprofits/edit.html.erb]: @ianroberts131 `text_field` will work fine here for email, website, and phone. There is also the choice of using more specific field types: `email_field`, `url_field`, and `phone_field`. Using these more specific field types when appropriate, can improve the user experience. For example, on touchscreen devices, an optimized keyboard is shown for the email fields where the "@" key is easy to access. See the various field types that are available here in the api docs: http://api.rubyonrails.org/classes/ActionView/Helpers/FormHelper.html

File: app/views/nonprofits/show.html.erb [hunk 3]
```diff
@@ -1,2 +1,19 @@
-<h1><%= "#{@nonprofit.name}" %></h1>
+<h1><%= @nonprofit.name %>
+  <% unless @nonprofit.managers.where(user_id: current_user.id).empty? %>
```
Comment (eliotsykes) [inline app/views/nonprofits/show.html.erb]: @ianroberts131 Line 2 works well, though its purpose is a little unclear on first reading it. 

The purpose here looks to be figuring out if the current user can edit the nonprofit.

If this is the case, consider writing a method on `User` that will reveal the purpose more clearly, say so the line could be rewritten to:

```
-  <% unless @nonprofit.managers.where(user_id: current_user.id).empty? %>
+  <% if current_user.can_edit?(@nonprofit) %>
```

File: app/views/users/registrations/edit.html.erb [hunk 4]
```diff
@@ -12,12 +12,12 @@
...
       <div class="form-group">
         <%= f.label :last_name %><br />
-        <%= f.text_field :last_name, required: true, autofocus: true, class: "form-control", placeholder: "Update last name" %>
+        <%= f.text_field :last_name, class: "form-control", placeholder: "Update last name" %>
       </div>
     
       <div class="form-group">
         <%= f.label :email %><br />
-        <%= f.email_field :email, required: true, autofocus: true, class: "form-control", placeholder: "Update email" %>
+        <%= f.email_field :email, class: "form-control", placeholder: "Update email" %>
```
Comment (eliotsykes) [inline app/views/users/registrations/edit.html.erb]: 👍 nice use of email_field

File: spec/features/update_users_spec.rb [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+    fill_in 'Email', with: user.email
+    fill_in 'Password', with: "password"
+    click_button 'Log in'
+    
+    click_link 'Edit profile'
+    
+    expect(page).to have_css('h2', text: 'Edit Profile')
+    fill_in 'First name', with: "New-name"
+    click_button 'Update profile'
+
```
Comment (eliotsykes) [inline spec/features/update_users_spec.rb]: @ianroberts131 If this is the test you mentioned, you might try debugging it by printing the page html and page text and seeing if there are any clues in the output as to why the profile is not getting updated:

```
# Try adding these lines after line 17, re-run the spec, and
# read the printed page content in the terminal:
puts "The page text is"
puts page.text
puts "The page HTML is"
puts page.html
```

**Comment by ianroberts131** [comment]
Eliot,

Thanks for this, I incorporated your changes. I knew what was wrong with the Edit User test - Devise was requiring the current password for any change and I didn't want that to be the case. I was able to search and find the answer on that one.

However, I'm not having an issue with the feature spec for updating a nonprofit. I get the following error:

```
Capybara::ElementNotFound:
       Unable to find select box "My Nonprofits"
```

I've done a lot of searching and can't figure out why this won't work. I'm going to move on, but maybe we can chat through this tomorrow.

**Comment by eliotsykes** [comment]
> Capybara::ElementNotFound: Unable to find select box "My Nonprofits"

Sure thing, we'll debug this one together, speak soon

File: app/views/users/registrations/edit.html.erb [hunk 6]
```diff
@@ -33,11 +33,6 @@
         <%= f.label :password_confirmation %><br />
         <%= f.password_field :password_confirmation, autocomplete: "off", required: false, class: "form-control", placeholder: "Confirm new password" %>
       </div>
-      
-      <div class="form-group">
-        <%= f.label :current_password %><br />
-        <%= f.password_field :current_password, hint: "we need your current password to confirm your changes", required: true, class: "form-control", placeholder: "Enter current password" %>
-      </div>
```
Comment (eliotsykes) [inline app/views/users/registrations/edit.html.erb]: @ianroberts131 Recommend keeping the current/old password requirement when a form allows changing a password. This stops an attacker who takes control of a user's session from locking that user out of their own account.

File: spec/features/update_nonprofits_spec.rb [hunk 7]
```diff
@@ -1,4 +1,32 @@
 require 'rails_helper'
 
 RSpec.feature "Update Nonprofits", type: :feature do
+  scenario 'successfully' do
+  
+  visit '/users/sign_up'
```
Comment (eliotsykes) [inline spec/features/update_nonprofits_spec.rb]: @ianroberts131 Check indentation for lines 6 to 29.

File: app/controllers/registrations_controller.rb [hunk 8]
```diff
@@ -4,6 +4,12 @@ def new
     super { |user| user.managers.build.build_nonprofit }
   end
 
+  protected
+  
+    def update_resource(resource, params)
+      resource.update_without_password(params)
+    end
```
Comment (eliotsykes) [inline app/controllers/registrations_controller.rb]: @ianroberts131 We can discuss the security implication of this change. See feedback at https://github.com/ianroberts131/coalesce-to-give/pull/22/commits/0c62d37e3cccd59f7ae4dbba79c380c39ac38823#r69263101

**Pull request merged**